File: Learning/src/Questions/BinaryTree/LevelOrderTraversal.java
package Questions.BinaryTree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LevelOrderTraversal {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }

    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (q.size() != 0) {
            int size = q.size();
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode curr = q.poll();
                temp.add(curr.val);
                if (curr.left != null) {
                    q.add(curr.left);
                }
                if (curr.right != null) {
                    q.add(curr.right);
                }

            }
            result.add(temp);

        }
        return result;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        TreeNode left = new TreeNode(9);
        TreeNode right = new TreeNode(20);
        TreeNode rightLeft = new TreeNode(15);
        TreeNode rightRight = new TreeNode(7);
        root.left = left;
        root.right = right;
        right.left = rightLeft;
        right.right = rightRight;
        LevelOrderTraversal levelOrderTraversal = new LevelOrderTraversal();
        List<List<Integer>> result = levelOrderTraversal.levelOrder(root);
        System.out.println("Level Order Traversal: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/LevelOrderTraversal.py
"""
# Question: Binary Tree Level Order Traversal
# Link: https://leetcode.com/problems/binary-tree-level-order-traversal/

# Time Complexity: O(N)
# Space Complexity: O(W) where W is max width

# Algorithm:
# 1. BFS traversal
# 2. Track levels
# 3. Build result list
"""

from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class LevelOrderTraversal:
    def level_order(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        result = []
        queue = deque([(root, 0)])

        while queue:
            level_size = len(queue)
            current_level = []

            for _ in range(level_size):
                node, level = queue.popleft()
                current_level.append(node.val)

                if node.left:
                    queue.append((node.left, level + 1))
                if node.right:
                    queue.append((node.right, level + 1))

            result.append(current_level)

        return result


def main():
    solution = LevelOrderTraversal()
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    print(solution.level_order(root))


if __name__ == "__main__":
    main()
